File: app/core/password_strength.py

import re

from dataclasses import dataclass

from enum import Enum

from typing import List, Tuple

class PasswordStrength(Enum):

    WEAK = 1

    FAIR = 2

    GOOD = 3

    STRONG = 4

@dataclass

class PasswordAnalysis:

    strength: PasswordStrength

    score: int

    label: str

    color: str

    feedback: List[str]

COMMON_PASSWORDS = {

    "password", "123456", "12345678", "qwerty", "abc123", "monkey",

    "1234567", "letmein", "trustno1", "dragon", "baseball", "iloveyou",

    "master", "sunshine", "ashley", "bailey", "shadow", "123123",

    "654321", "superman", "qazwsx", "michael", "football", "password1",

    "password123", "welcome", "jesus", "ninja", "mustang", "password12",

}

KEYBOARD_PATTERNS = [

    "qwerty", "qwertz", "azerty", "asdfgh", "zxcvbn",

    "1234567890", "0987654321", "qazwsx", "1qaz2wsx",

]

def _has_lowercase(password: str) -> bool:

    return bool(re.search(r'[a-z]', password))

def _has_uppercase(password: str) -> bool:

    return bool(re.search(r'[A-Z]', password))

def _has_digits(password: str) -> bool:

    return bool(re.search(r'\d', password))

def _has_special_chars(password: str) -> bool:

    return bool(re.search(r'[!@#$%^&*()_+\-=\[\]{}|;:\'",.<>?/`~\\]', password))

def _has_repeated_chars(password: str, max_repeat: int = 3) -> bool:

    pattern = r'(.)\1{' + str(max_repeat - 1) + r',}'

    return bool(re.search(pattern, password))

def _has_sequential_chars(password: str) -> bool:

    sequences = [

        "abcdefghijklmnopqrstuvwxyz",

        "0123456789",

        "qwertyuiop",

        "asdfghjkl",

        "zxcvbnm",

    ]

    password_lower = password.lower()

    for seq in sequences:

        for i in range(len(seq) - 2):

            if seq[i:i+3] in password_lower:

                return True

            if seq[i:i+3][::-1] in password_lower:

                return True

    return False

def _is_common_password(password: str) -> bool:

    return password.lower() in COMMON_PASSWORDS

def _has_keyboard_pattern(password: str) -> bool:

    password_lower = password.lower()

    for pattern in KEYBOARD_PATTERNS:

        if pattern in password_lower:

            return True

    return False

def _calculate_entropy_score(password: str) -> float:

    charset_size = 0

    if _has_lowercase(password):

        charset_size += 26

    if _has_uppercase(password):

        charset_size += 26

    if _has_digits(password):

        charset_size += 10

    if _has_special_chars(password):

        charset_size += 32

    if charset_size == 0:

        return 0

    import math

    entropy = len(password) * math.log2(charset_size)

    return min(100, (entropy / 128) * 100)

def analyze_password(password: str) -> PasswordAnalysis:

    if not password:

        return PasswordAnalysis(

            strength=PasswordStrength.WEAK,

            score=0,

            label="EMPTY",

            color="#ef4444",

            feedback=["Enter a password"]

        )

    score = 0

    feedback = []

    length = len(password)

    if length < 6:

        score += 5

        feedback.append("Use at least 8 characters")

    elif length < 8:

        score += 10

        feedback.append("Consider using 12+ characters")

    elif length < 12:

        score += 20

        feedback.append("Great! Consider 16+ for maximum security")

    elif length < 16:

        score += 25

    else:

        score += 30

    has_lower = _has_lowercase(password)

    has_upper = _has_uppercase(password)

    has_digit = _has_digits(password)

    has_special = _has_special_chars(password)

    variety_count = sum([has_lower, has_upper, has_digit, has_special])

    score += variety_count * 10

    if not has_lower:

        feedback.append("Add lowercase letters")

    if not has_upper:

        feedback.append("Add uppercase letters")

    if not has_digit:

        feedback.append("Add numbers")

    if not has_special:

        feedback.append("Add special characters (!@#$%...)")

    if _is_common_password(password):

        score -= 30

        feedback.insert(0, "Avoid common passwords")

    if _has_repeated_chars(password):

        score -= 10

        feedback.append("Avoid repeated characters (aaa, 111)")

    if _has_sequential_chars(password):

        score -= 10

        feedback.append("Avoid sequential patterns (abc, 123)")

    if _has_keyboard_pattern(password):

        score -= 10

        feedback.append("Avoid keyboard patterns (qwerty)")

    entropy_bonus = _calculate_entropy_score(password) * 0.2

    score += int(entropy_bonus)

    score = max(0, min(100, score))

    if score < 30:

        strength = PasswordStrength.WEAK

        label = "WEAK"

        color = "#ef4444"

    elif score < 50:

        strength = PasswordStrength.FAIR

        label = "FAIR"

        color = "#f97316"

    elif score < 75:

        strength = PasswordStrength.GOOD

        label = "GOOD"

        color = "#eab308"

    else:

        strength = PasswordStrength.STRONG

        label = "STRONG"

        color = "#22c55e"

        feedback = ["Excellent password!"] if not feedback else feedback[:1]

    feedback = feedback[:3]

    return PasswordAnalysis(

        strength=strength,

        score=score,

        label=label,

        color=color,

        feedback=feedback

    )

def get_strength_bar_segments(analysis: PasswordAnalysis) -> List[Tuple[str, bool]]:

    strength_level = analysis.strength.value

    colors = [

        "#ef4444",

        "#f97316",

        "#eab308",

        "#22c55e",

    ]

    segments = []

    for i in range(4):

        is_filled = i < strength_level

        color = analysis.color if is_filled else "#374151"

        segments.append((color, is_filled))

    return segments

File: extension/password_strength.js

const COMMON_PASSWORDS = new Set([
  "password",
  "123456",
  "12345678",
  "qwerty",
  "abc123",
  "monkey",
  "1234567",
  "letmein",
  "trustno1",
  "dragon",
  "baseball",
  "iloveyou",
  "master",
  "sunshine",
  "ashley",
  "bailey",
  "shadow",
  "123123",
  "654321",
  "superman",
  "qazwsx",
  "michael",
  "football",
  "password1",
  "password123",
  "welcome",
  "jesus",
  "ninja",
  "mustang",
  "password12",
]);

const KEYBOARD_PATTERNS = [
  "qwerty",
  "qwertz",
  "azerty",
  "asdfgh",
  "zxcvbn",
  "1234567890",
  "0987654321",
  "qazwsx",
  "1qaz2wsx",
];

const PasswordStrength = {
  WEAK: 1,
  FAIR: 2,
  GOOD: 3,
  STRONG: 4,
};

function hasLowercase(password) {
  return /[a-z]/.test(password);
}

function hasUppercase(password) {
  return /[A-Z]/.test(password);
}

function hasDigits(password) {
  return /\d/.test(password);
}

function hasSpecialChars(password) {
  return /[!@#$%^&*()_+\-=\[\]{}|;:'",.<>?/`~\\]/.test(password);
}

function hasRepeatedChars(password, maxRepeat = 3) {
  const pattern = new RegExp(`(.)\\1{${maxRepeat - 1},}`);
  return pattern.test(password);
}

function hasSequentialChars(password) {
  const sequences = [
    "abcdefghijklmnopqrstuvwxyz",
    "0123456789",
    "qwertyuiop",
    "asdfghjkl",
    "zxcvbnm",
  ];
  const passwordLower = password.toLowerCase();

  for (const seq of sequences) {
    for (let i = 0; i < seq.length - 2; i++) {
      const sub = seq.substring(i, i + 3);
      if (passwordLower.includes(sub)) return true;

      const subRev = sub.split("").reverse().join("");
      if (passwordLower.includes(subRev)) return true;
    }
  }
  return false;
}

function isCommonPassword(password) {
  return COMMON_PASSWORDS.has(password.toLowerCase());
}

function hasKeyboardPattern(password) {
  const passwordLower = password.toLowerCase();
  return KEYBOARD_PATTERNS.some((pattern) => passwordLower.includes(pattern));
}

function calculateEntropyScore(password) {
  let charsetSize = 0;
  if (hasLowercase(password)) charsetSize += 26;
  if (hasUppercase(password)) charsetSize += 26;
  if (hasDigits(password)) charsetSize += 10;
  if (hasSpecialChars(password)) charsetSize += 32;

  if (charsetSize === 0) return 0;

  const entropy = password.length * Math.log2(charsetSize);
  return Math.min(100, (entropy / 128) * 100);
}

function analyzePassword(password) {
  if (!password) {
    return {
      strength: PasswordStrength.WEAK,
      score: 0,
      label: "EMPTY",
      color: "var(--vk-accent-red)",
      feedback: ["Enter a password"],
    };
  }

  let score = 0;
  let feedback = [];
  const length = password.length;
  if (length < 6) {
    score += 5;
    feedback.push("Use at least 8 characters");
  } else if (length < 8) {
    score += 10;
    feedback.push("Consider using 12+ characters");
  } else if (length < 12) {
    score += 20;
    feedback.push("Great! Consider 16+ for maximum security");
  } else if (length < 16) {
    score += 25;
  } else {
    score += 30;
  }
  const hasLower = hasLowercase(password);
  const hasUpper = hasUppercase(password);
  const hasDigit = hasDigits(password);
  const hasSpecial = hasSpecialChars(password);

  const varietyCount = [hasLower, hasUpper, hasDigit, hasSpecial].filter(
    Boolean,
  ).length;
  score += varietyCount * 10;

  if (!hasLower) feedback.push("Add lowercase letters");
  if (!hasUpper) feedback.push("Add uppercase letters");
  if (!hasDigit) feedback.push("Add numbers");
  if (!hasSpecial) feedback.push("Add special characters");
  if (isCommonPassword(password)) {
    score -= 30;
    feedback.unshift("Avoid common passwords");
  }

  if (hasRepeatedChars(password)) {
    score -= 10;
    feedback.push("Avoid repeated characters");
  }

  if (hasSequentialChars(password)) {
    score -= 10;
    feedback.push("Avoid sequential patterns");
  }

  if (hasKeyboardPattern(password)) {
    score -= 10;
    feedback.push("Avoid keyboard patterns");
  }
  const entropyBonus = calculateEntropyScore(password) * 0.2;
  score += Math.floor(entropyBonus);
  score = Math.max(0, Math.min(100, score));
  let strength, label, color;

  if (score < 30) {
    strength = PasswordStrength.WEAK;
    label = "Weak";
    color = "var(--vk-accent-red)";
  } else if (score < 50) {
    strength = PasswordStrength.FAIR;
    label = "Fair";
    color = "var(--vk-accent-warning)";
  } else if (score < 75) {
    strength = PasswordStrength.GOOD;
    label = "Good";
    color = "#eab308";
  } else {
    strength = PasswordStrength.STRONG;
    label = "Fantastic";
    color = "var(--vk-accent-green)";
    if (feedback.length === 0) feedback = ["Excellent password!"];
  }

  return {
    strength,
    score,
    label,
    color,
    feedback: feedback.slice(0, 3),
  };
}